### ISSUE TITLE
url_launcher can not launch URLs from a ListTiles onLongPress() method on mobile Safari browsers

### ISSUE THREAD

**Issue by mbalmer** [flutter/flutter#117775]
## Steps to Reproduce

1. Execute the code sample and view it on a mobile Safari browser running on iOS 16.2 or iPadOS 16.2
2. You can view the sample code below at https://scoremaker.ch/launcher_issue (caveat: this URL will eventually go away)
3. Long press on the ListTile, nothing happens in mobile Safari, other browsers take you to the website as expected.

**Expected results:** A long press on the ListTile should take you to https://flutter.dev

**Actual results:** Nothing happens

<details>
<summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';

import 'package:url_launcher/url_launcher.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'URL Launcher Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'URL Launcher Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  void _launchMyURL(String url) async {
    await launchUrl(Uri.parse(url));
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            const Text('Long press the list tile to go to https://flutter.dev:'),
            Card(
                child: ListTile(
                    title: Text('https://flutter.dev',
                        style: TextStyle(fontSize: 18.0)),
                    onLongPress: () => _launchMyURL('https://flutter.dev'),
                    trailing: Icon(Icons.link)))
          ],
        ),
      ),
    );
  }
}

```

</details>

<details>

**Comment by exaby73** [comment]
## Triage report

I can reproduce this issue on an iOS 16.2 simulator on Safari.

<details><summary>Code Sample (Same as OP)</summary>

```dart
import 'package:flutter/material.dart';

import 'package:url_launcher/url_launcher.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'URL Launcher Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'URL Launcher Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  void _launchMyURL(String url) async {
    await launchUrl(Uri.parse(url));
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            const Text('Long press the list tile to go to https://flutter.dev:'),
            Card(
                child: ListTile(
                    title: Text('https://flutter.dev',
                        style: TextStyle(fontSize: 18.0)),
                    onLongPress: () => _launchMyURL('https://flutter.dev'),
                    trailing: Icon(Icons.link)))
          ],
        ),
      ),
    );
  }
}
```

</details>

**Comment by ditman** [comment]
Quick test: If the `onLongPress` function is `() { print('onLongPress'); }` (or something like that, that doesn't attempt to open an URL), does it work?

We're trying to figure out if this is an issue with the onLongPress event itself not being handled correctly (engine or framework issue), or maybe "long press" not being a valid user interaction for the launch URL implementation (plugin issue).

**Comment by mbalmer** [comment]
> Quick test: If the onLongPress function is () { print('onLongPress'); } (or something like that, that doesn't attempt to open an URL), does it work?
> We're trying to figure out if this is an issue with the onLongPress event itself not being handled correctly (engine or framework issue), or maybe "long press" not being a valid user interaction for the launch URL implementation (plugin issue).
> 
As described, the onLongPress event is handled correctly.  It is launchUrl() that does nothing.

**Comment by ditman** [comment]
Without actually having had time to debug the code; if `_launchMyURL` is running, but not launching the URL, I suspect you're hitting this feature of the web platform:

* https://developer.mozilla.org/en-US/docs/Web/Security/User_activation

This needs more triage, but it is likely that when Flutter dispatches the "long press" event, the user interaction is not "trusted" yet, so the browser denies the navigation (as if it were a popup blocker).

(I'm assuming the "normal" `onPress` event works?)

**Comment by mbalmer** [comment]
> Without actually having had time to debug the code; if _launchMyURL is running, but not launching the URL, I suspect you're hitting this feature of the web platform:
> https://developer.mozilla.org/en-US/docs/Web/Security/User_activation
> This needs more triage, but it is likely that when Flutter dispatches the "long press" event, the user interaction is not "trusted" yet, so the browser denies the navigation (as if it were a popup blocker).
> 
> (I'm assuming the "normal" onPress event works?)
> 

Yes, only onLongPress is affected.

Also, I have a new finding which further narrows down the issue: The URL is only then not launched when you long press on the glass/touchscreen, if you use a mouse with your iPad, the URL is launched.

**Comment by exaby73** [comment]
I can confirm the behavior @mbalmer is describing. Showing a snackbar for example on `onLongPress` works. Just not the URL launch

**Comment by mbalmer** [comment]
> 
>> Without actually having had time to debug the code; if _launchMyURL is running, but not launching the URL, I suspect you're hitting this feature of the web platform:
>> https://developer.mozilla.org/en-US/docs/Web/Security/User_activation
>> This needs more triage, but it is likely that when Flutter dispatches the "long press" event, the user interaction is not "trusted" yet, so the browser denies the navigation (as if it were a popup blocker).
>> 
>> (I'm assuming the "normal" onPress event works?)
>> 
> 
> Yes, only onLongPress is affected.
> 
> Also, I have a new finding which further narrows down the issue: The URL is only then not launched when you long press on the glass/touchscreen, if you use a mouse with your iPad, the URL is launched.

Further findings:

The issue also happens on Android using the Chrome Browser, Chrome, however, displays a message: „Pop-up blocked“.  Edge on Android works as expected.

As a cross check I used a Firefox Browser on a Linux (openSUSE Leap 15.4) machine that has a touchscreen.  It works as it should, i.e. a long press on the glass takes you to the website.

My (wild) guess is, that onTap() sets user interaction to trusted, while onLongPress() does not, leaving it at the browsers discretion whether to open the URL or not.

**Comment by stuartmorgan** [comment]
Marking P6 as it seems unlikely we can control this; web folks please feel free to either change the priority or close this as unactionable.

**Comment by mbalmer** [comment]
Is this way you deal with bugs? Simply close them?Enough data has been collected to find the root cause and fix it, imo.Am 03.01.2023 um 17:28 schrieb stuartmorgan ***@***.***>:﻿
Marking P6 as it seems unlikely we can control this; web folks please feel free to either change the priority or close this as unactionable.

**Comment by ditman** [comment]
> My (wild) guess is, that onTap() sets user interaction to trusted, while onLongPress() does not, leaving it at the browsers discretion whether to open the URL or not.

@mbalmer The [`isTrusted` boolean](https://developer.mozilla.org/en-US/docs/Web/API/Event/isTrusted) property of events in JS is a read-only property **set by the browser** when the event was dispatched by the browser itself (and not synthesized or modified by a script).

[User activation](https://developer.mozilla.org/en-US/docs/Web/Security/User_activation) requires an event to be trusted AND of certain types to work, relevant to this case:

* `mousedown` / `pointerdown` (**when using a mouse**)
* `touchend` / `pointerup` (when using touch)

Long-press events fire a few hundred milliseconds (about 500ms) after the _pointing-device_`down` event, but **before the _pointing-device_`up` event**.

Unfortunately, **Long-press is not a "native" event in web browsers** that we can attach a listener to (on mobile, it's normally reserved to interactions like "right click" / "download photo" / "open link in new tab..."), so it's always "synthesized" by the framework (after a few hundred ms of seeing a _pointing-device_`down` event).

In conclusion/TL;DR:

* When `longTap` is synthesized from a `pointerdown` event triggered **by a mouse**, user activation will be enabled, and links will work.
* When `longTap` is synthesized from a `pointerdown` event triggered **by touch**, user activation will **NOT** be enabled, and links will **not** work (popup blockers will do their thing, etc...)

Unfortunately, and as @stuartmorgan suspected, this is something we cannot control, since it seems like a fundamental limitation of how browsers work. I'm sorry that this is not the answer you expected @mbalmer!

(Labeling as `documentation` so we can come back to this when writing a document about this types of differences in the web platform. This is especially painful because of the difference between *mouse* and *touch* devices!)
